### PR TITLE
Save and Load Data Preparation Pipelines and Trained Models How-To

### PR DESCRIPTION
**Pull request by luisquintanilla**
Updated Save and Load Data Preparation Pipelines and Trained Models How-To

[Internal Review Link](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/consuming-model-ml-net?branch=pr-en-us-11990)

[AB#1484938](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1484938)

### CODE REVIEW

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 1]
```diff
@@ -1,32 +1,125 @@
...
+Learn how to save and load data preparation pipelines and trained models in your application. Although this sample uses a regression model, the same process applies to other ML.NET algorithms.
 
-When the model metrics look good to you, it's time to 'operationalize' the model. The `model` object you built can be consumed, persisted, and reused in different environments, applying the same steps that it 'learned' during training.
-
-To save the model to a file, and reload it (potentially in a different context), use the following example:
+Throughout the model building process, a model lives in memory and is accessible throughout the application's lifecycle. However, once the application stops running, if the model is not saved somewhere locally or remotely, it is no longer accessible. Typically models are used at some point after training either for inference or re-training. Therefore, it is important to store the model as well as its data preparation pipeline. Using a data preparation and model training pipeline like the one detailed below, save and load these components using the steps described in subsequent sections of this document.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: Use contractions. However, once the application stops running, if the model **isn't** saved somewhere locally or remotely, **it's** no longer accessible. Therefore, **it's** important to store the model as well as its data preparation pipeline.

Comment (JRAlexander) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: Maybe something like: "Save and load these components using the steps described in subsequent sections of this document when using a data preparation and model training pipeline like the one detailed below."?

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 2]
```diff
@@ -1,32 +1,125 @@
...
+// Define StochasticDualCoordinateAscent regression algorithm estimator
+var sdcaEstimator = mlContext.Regression.Trainers.Sdca();
+
+// Train model
+var trainedModel = sdcaEstimator.Fit(transformedData);
+```
+
+## Save Data Preparation Pipelines and Model Locally
+
+When saving a data preparation pipeline or trained model, you will need two things:
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: I would remove the 'will'.

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 3]
```diff
@@ -1,32 +1,125 @@
...
+
+1. The [`ITransformer`](xref:Microsoft.ML.ITransformer) of the data preparation pipeline or model.
+2. The [`DataViewSchema`](xref:Microsoft.ML.DataViewSchema) of the [`ITransformer`](xref:Microsoft.ML.ITransformer)'s expected input.
+
+Before providing input data to a model, it needs to be pre-processed. Therefore, it is important to save the data preparation steps along with the model to get the data into the format expected by the model.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: " Therefore, **it's** important "

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 4]
```diff
@@ -1,32 +1,125 @@
...
-using (var stream = File.OpenRead(modelPath))
-    loadedModel = mlContext.Model.Load(stream);
+```csharp
+// Save Trained Model
+mlContext.Model.Save(trainedModel, transformedData.Schema, "model.zip");
+```
+
+## Load Data Preparation Pipelines and Model Stored Locally
+
+Models stored on disk can be used in other processes or applications like ASP.NET Core and Serverless Web Applications. See [Use ML.NET in Web API](./serve-model-web-api-ml-net.md) and [Deploy ML.NET Serverless Web App](./serve-model-serverless-azure-functions-ml-net.md) how-to articles to learn more.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: Could you just say "Stored models" without "on disk"?

Comment (JRAlexander) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: ASP.NET Core and Serverless Web Applications should be code fenced

**Review by JRAlexander — CHANGES_REQUESTED** [review]
Really good! Just a few items.

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 1 above)
Comment (natke) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: I wonder if it would be better to use model generically for the most of the doc, and then at the bottom, make the distinction the model and the data preparation pipeline.

So .. structure would be:

Title: Save and Load Models

h2: Save mode locally
h2: Load model from local storage
h2: Load model from remote storage
h2: Save and load data preparation pipeline separately

**Review by natke — COMMENTED** [review]
Looks good. A couple of comments.

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 4 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 5]
```diff
@@ -1,32 +1,125 @@
...
 #Customer intent: As a developer, I want to use ML.NET to consume my trained and evaluated machine learning model in my applications.
 ---
-# Operationalize a trained machine learning model in apps - ML.NET
 
-> [!NOTE]
-> This topic refers to ML.NET, which is currently in Preview, and material may be subject to change. For more information, visit [the ML.NET introduction](https://www.microsoft.com/net/learn/apps/machine-learning-and-ai/ml-dotnet).
+# How-To: Save and Load Data Preparation Pipelines and Trained Models
 
-This how-to and related sample are currently using **ML.NET version 0.10**. For more information, see the release notes at the [dotnet/machinelearning github repo](https://github.com/dotnet/machinelearning/tree/master/docs/release-notes)
+Learn how to save and load data preparation pipelines and trained models in your application. Although this sample uses a regression model, the same process applies to other ML.NET algorithms.
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 3 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 2 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 1 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 6]
```diff
@@ -1,32 +1,160 @@
...
+Learn how to save and load trained models in your application. 
 
-When the model metrics look good to you, it's time to 'operationalize' the model. The `model` object you built can be consumed, persisted, and reused in different environments, applying the same steps that it 'learned' during training.
+Throughout the model building process, a model lives in memory and is accessible throughout the application's lifecycle. However, once the application stops running, if the model is not saved somewhere locally or remotely, it's no longer accessible. Typically models are used at some point after training in other applications either for inference or re-training. Therefore, it's important to store the model. Save and load models using the steps described in subsequent sections of this document when using data preparation and model training pipelines like the one detailed below. Although this sample uses a linear regression model, the same process applies to other ML.NET algorithms.
 
-To save the model to a file, and reload it (potentially in a different context), use the following example:
+```csharp
+HousingData[] housingData = new HousingData[]
+            {
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: >           { [](start = 1, length = 12)

indentation

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 7]
```diff
@@ -1,32 +1,160 @@
...
+        .Append(mlContext.Regression.Trainers.Sdca());
+
+// Train model
+ITransformer trainedModel = pipelineEstimator.Fit(data);
+
+// Save model
+mlContext.Model.Save(trainedModel, data.Schema, "model.zip");
+```
+
+Because most models and data preparation pipelines inherit from the same set of classes, the save and load operations for these components is the same. Depending on your use case, you can either combine the data preparation pipeline and model into a single [`EstimatorChain`](xref:Microsoft.ML.Data.TransformerChain%601) which would output a single [`ITransformer`](xref:Microsoft.ML.ITransformer) or separate them thus creating separate `ITransformers` for each.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: >save and load operations for these components is the same [](start = 93, length = 57)

maybe be more concrete: the save and load api signature is similar.

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 8]
```diff
@@ -1,32 +1,160 @@
...
-    mlContext.Model.Save(model, stream);
+    Stream modelFile = await client.GetStreamAsync("<YOUR-REMOTE-FILE-LOCATION>");
+
+    trainedModel = mlContext.Model.Load(modelFile, out modelSchema);
 }
+```
+
+## Separate Data Preparation Pipelines and Models
+
+When working with separate data preparation pipelines and models, the same operations apply, except they both now components need to be saved and loaded simultaneously.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: >they both now components need [](start = 100, length = 29)

this gives me the impression that the components of the pipelines need to be separately saved and loaded. 

Maybe rephrase to make it more clear that it is the pipelines that need to be saved/loaded.

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md [hunk 9]
```diff
@@ -1,32 +1,160 @@
...
+author: luisquintanilla
+ms.author: luquinta
+ms.custom: mvc, how-to
 #Customer intent: As a developer, I want to use ML.NET to consume my trained and evaluated machine learning model in my applications.
 ---
-# Operationalize a trained machine learning model in apps 
 
-> [!NOTE]
-> This topic refers to ML.NET, which is currently in Preview, and material may be subject to change. For more information, visit [the ML.NET introduction](https://www.microsoft.com/net/learn/apps/machine-learning-and-ai/ml-dotnet).
+# How-To: Save and load trained models
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: > How-To: Save and load trained models [](start = 1, length = 37)

as a general feedback, the users don;t necessary need to separate the data prep pipeline, from the model pipeline. The only reason to do that, is to avoid the pain with casting, if they want to access the model parameters, and inspect the model. 

If all they want to do is inference, they are fine to use one pipeline. 

Maybe we can make a note of this, and have one example where the two pipelines are merged.

**Review by sfilipi — APPROVED** [review]
:shipit:

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 6 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 7 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 8 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

File: docs/machine-learning/how-to-guides/consuming-model-ml-net.md (hunk 9 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/consuming-model-ml-net.md]: #resolved

**Review by JRAlexander — APPROVED** [review]

**Comment by JRAlexander** [comment]
Nice work, @luisquintanilla!

**Review by natke — APPROVED** [review]
Much improved!
